### PR TITLE
Add separable nd-filters and sliding window

### PR DESCRIPTION
**Pull request by fzimmermann89**
some utilities: 
uniform filter,  gaussian filter, sliding windows views

### CODE REVIEW

**Comment by fzimmermann89** [comment]
I am not sure setting the windows stride greater than 1 is always correct, thus the warning
This is like unfold for arbitrary number of dimensions, similiar to using numpys stride tricks.

File: src/mrpro/utils/filters.py [hunk 1]
```diff
@@ -33,22 +35,116 @@ def spatial_uniform_filter_3d(
...
+        List of 1D kernels to apply to the tensor x.
+    axis
+        axes to filter over. Must have the same length as kernels.
+    """
+    if len(axis) != len(kernels):
+        raise ValueError('Must provide matching length kernels and axis arguments. ')
+    if len(axis) > x.ndim:
+        raise ValueError('Too many axes provided')
+    for kernel, ax in zip(kernels, axis):
+        if x.is_complex() and not kernel.is_complex():
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py:82]: needs comment

File: src/mrpro/utils/filters.py [hunk 2]
```diff
@@ -33,22 +35,116 @@ def spatial_uniform_filter_3d(
...
+            kernel = kernel + 0.0j
+        elif kernel.is_complex() and not x.is_complex():
+            x = x + 0.0j
+        # moveaxis is not implemented for batched tensors, so vmap would fail
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: instead we use permute

File: src/mrpro/utils/filters.py [hunk 3]
```diff
@@ -33,22 +35,116 @@ def spatial_uniform_filter_3d(
...
+    axis
+        Axis or axes to filter over. If None, filters over all axes.
+    truncate
+        Truncate the filter at this many standard deviations.
+    """
+    if axis is None:
+        axis = tuple(range(x.ndim))
+    elif isinstance(axis, int):
+        axis = (axis,)
+    sigmas = torch.tensor(sigmas) if np.iterable(sigmas) else torch.tensor([sigmas] * len(axis))
```
Comment (fzimmermann89) [inline src/mrpro/utils/filters.py]: will cause warning if sigmas is tensor.

File: src/mrpro/utils/window.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import warnings
+
+import torch
+
+_iterables = (torch.Tensor, tuple, list)
+
+
+def sliding_window(
+    x: torch.Tensor,
+    window_shape: int | tuple[int, ...],
```
Comment (fzimmermann89) [inline src/mrpro/utils/window.py]: list or tuple?

File: src/mrpro/utils/window.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    if torch.any(window_shape_arr < 0):
+        raise ValueError('window_shape cannot contain negative values')
+    if torch.any(strides_arr < 0):
+        raise ValueError('strides cannot contain negative values')
+    if len(window_shape) != len(axis):
+        raise ValueError('Must provide matching length window_shape and axis arguments. ')
+    if len(strides) != len(axis):
+        raise ValueError('Must provide matching length strides and axis arguments.')
+
+    out_strides = torch.tensor([x.stride(i) for i in range(x.ndim)] + [x.stride(ax) for ax in axis])
```
Comment (fzimmermann89) [inline src/mrpro/utils/window.py]: needs a comment to explain the logic!

File: src/mrpro/utils/filters.py [hunk 6]
```diff
@@ -33,22 +37,128 @@ def spatial_uniform_filter_3d(
         Width of 3D the filter as SpatialDimension(z, y, x) or tuple(z, y, x).
```
Comment (schuenke) [inline src/mrpro/utils/filters.py]: remove "the"

File: src/mrpro/utils/filters.py [hunk 7]
```diff
@@ -33,22 +37,128 @@ def spatial_uniform_filter_3d(
...
         The filter width is clipped to the data shape.
     """
-
-    # Create a box-shaped filter kernel
     match filter_width:
```
Comment (schuenke) [inline src/mrpro/utils/filters.py:44]: should we allow a int here assuming that a single int means the same width in all 3 dimensions?

File: tests/utils/test_window.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    no error handling, expanded parmeters, no fancy indexing, copies
+    """
+    for ax, ws, st in zip(axis, window_shape, stride):
+        data = data.unfold(ax, ws, st)
+    return data
+
+
+@pytest.mark.filterwarnings('ignore:strides other than 1 are not fully supported')
+@pytest.mark.parametrize(
+    'shape, window_shape, axis, stride',
```
Comment (schuenke) [inline tests/utils/test_window.py]: should be a tuple of strings to pass ruff checks after merge of #193

File: tests/utils/test_filters.py [hunk 9]
```diff
@@ -29,19 +31,101 @@ def data():
...
 def test_spatial_uniform_filter_wrong_width(data):
     """Test spatial_uniform_filter_3d with wrong width."""
 
     with pytest.raises(ValueError, match='Invalid filter width'):
         spatial_uniform_filter_3d(data, (3, 3))
+
+
+def test_gaussian_filter_int_axis(data):
+    result = gaussian_filter(data, 0.5, -1)
+    assert result.nonzero() == torch.tensor(
```
Comment (schuenke) [inline tests/utils/test_filters.py]: should be assert torch.equal(result, tensor), no ?!

File: tests/utils/test_filters.py [hunk 10]
```diff
@@ -29,19 +31,101 @@ def data():
...
+    torch.testing.assert_close(result.sum(), torch.tensor(1.0), atol=1e-3, rtol=0)
+
+
+def test_gaussian_filter_two_sigmas(data):
+    result = gaussian_filter(data, (0.2, 0.5), (-1, 3))
+    assert result.shape == data.shape
+    torch.testing.assert_close(result.sum(), torch.tensor(1.0), atol=1e-3, rtol=0)
+
+
+def test_gaussian_filter_noaxis(data):
```
Comment (schuenke) [inline tests/utils/test_filters.py:82]: is it worth adding a test to ensure that the results along the different axis match? e.g. result[0,0,:,2,2] == result[0,0,2,:,2] == ... ?

File: src/mrpro/utils/window.py
Comment (schuenke) [inline src/mrpro/utils/window.py:1]: rename window.py -> sliding_window.py?

File: tests/utils/test_filters.py
Comment (schuenke) [inline tests/utils/test_filters.py:1]: maybe add a test for the truncate arg of gaussian_filter?

File: src/mrpro/utils/window.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    if torch.any(window_shape_tensor < 0):
+        raise ValueError('window_shape cannot contain negative values')
+    if torch.any(strides_tensor < 0):
+        # this is a pytorch limitation. python api standard should allow negative strides
+        raise ValueError('strides cannot contain negative values')
+    if len(window_shape) != len(axis):
+        raise ValueError('Must provide matching length window_shape and axis arguments. ')
+    if len(strides) != len(axis):
+        raise ValueError('Must provide matching length strides and axis arguments.')
+    # out_stides should be the original strides, but for sliding windows axis the stride should be increased
```
Comment (schuenke) [inline src/mrpro/utils/window.py]: out_stides  -> out_strides

File: src/mrpro/utils/filters.py (hunk 7 above)
Comment (fzimmermann89) [reply src/mrpro/utils/filters.py:44]: I want to remove this function completely.

**Comment by fzimmermann89** [comment]
@schuenke @JoHa0811

File: src/mrpro/utils/filters.py (hunk 6 above)
Comment (schuenke) [reply src/mrpro/utils/filters.py]: Should be "Width of the 3D filter as SpatialDimension(z, y, x), tuple(z, y, x) or single int."

**Review by schuenke — APPROVED** [review]

File: tests/utils/test_filters.py
Comment (fzimmermann89) [reply tests/utils/test_filters.py:1]: yes, that would be nice, but i cant think of a useful test invariant....
